scripts/bootstrap.sh

```diff
@@ -66,35 +68,79 @@ else
         # libpython2.7-stdlib : Breaks: python-urllib3 (< 1.9.1-3) but 1.7.1-1ubuntu4 is to be installed
         # due to the previous PPA
         add-apt-repository -y ppa:ross-kallisti/python-urllib3
-        upgrade_python=true
+        upgrade_python2=true
     fi
 
     # if flag present, upgrade python
     if [ -f /root/upgrade-python.flag ]; then
-        upgrade_python=true
+        upgrade_python2=true
+    fi
+fi
+
+if ! command -v python3; then
+    # python3 not found
+    upgrade_python3=true
+else
+    # python 3 found but have our other py3 dependencies been installed?
+    if ! grep "installed/upgraded python3" /root/events.log; then
+        upgrade_python3=true
     fi
 fi
 
 if ! dpkg -l git; then
+    # git not found
     install_git=true
 fi
 
-if ($upgrade_python || $install_git); then
+if ($upgrade_python2 || $upgrade_python3 || $install_git); then
     apt-get update -y
 fi
 
-if $upgrade_python; then
-    apt-get install python2.7 python2.7-dev -y
-    # virtual envs have to be recreated
-    find /srv /opt -depth -type d -name venv -exec rm -rf "{}" \;
 
-    # install/upgrade pip+setuptools
-    apt-get install python-pip python-setuptools --no-install-recommends -y
-    python2.7 -m pip install pip setuptools --upgrade
+# flag we can toggle to disable installation of python2 and python2 libs.
+# set to `false` once all formulas and formula code has been updated.
+elife_depends_on_python2=true
+
+if $upgrade_python2; then
+
+    if $elife_depends_on_python2; then
+        echo "eLife still has formulas that depend on Python2!"
+        apt-get install python2.7 python2.7-dev -y
 
+        # virtualenvs have to be recreated
+        find /srv /opt -depth -type d -name venv -exec rm -rf "{}" \;
+
+        # install/upgrade pip+setuptools
+        apt-get install python-pip python-setuptools --no-install-recommends -y
+        python2.7 -m pip install pip setuptools --upgrade
+    fi
+
+    # remove flag, if it exists
     rm -f /root/upgrade-python.flag
 fi
 
+if $upgrade_python3; then
+    apt-get install python3 python3-dev python3-pip python3-setuptools -y --no-install-recommends
+    python3 -m pip install pip setuptools --upgrade
+
+    # virtualenvs have to be recreated
+    find /srv /opt -depth -type d -name venv -exec rm -rf "{}" \;
+
+    # some libraries need to be installed *before* calling Salt
+    python3 -m pip install "docker[tls]==4.1.0"
+
+    # record an entry about when python3 was installed/upgraded
+    # presence of this entry is used to skip this section in future, unless forced with a flag
+    if [ -f /root/upgrade-python3.flag ]; then
+        echo "$(date -I) -- installed/upgraded python3 (forced)" >> /root/events.log;
+    else
+        echo "$(date -I) -- installed/upgraded python3" >> /root/events.log;
+    fi
+
+    # remove 'force upgrade' flag, if it exists
+    rm -f /root/upgrade-python3.flag
```

Comment:
could be inside the `then`?

Reply:
It could be. The purpose of that flag is to force an upgrade of python2. It's unlikely, but if we set it and `elife_depends_on_python2` was `false`, the flag would hang around forever.

That whole block will gone soon enough. It's barely worth thinking about.